evo1/atb/base.py

```diff
@@ -87,94 +21,98 @@ def _tap_confirm() -> None:
-        rng = EvolandRNG().get_rng()
-        rand_float = rng.rand_float()
-        attack = attacker.attack
-        defense = defender.defense
-        return int(attack + (0.5 * attack * rand_float) - defense + 0.5)
-
-    def predict_hit(self, attacker: BattleEntity, defender: BattleEntity) -> bool:
-        rng = EvolandRNG().get_rng()
-        # accuracy = 10  # TODO: from attacker
-        # evade = 10  # TODO: from defender
-        miss_chance = 30  # TODO: Calc
-        roll = (rng.rand_int() & 0x3FFFFFFF) % 100
-        return roll >= miss_chance
-
     # TODO: Actual combat logic
     # TODO: Overload with more complex
     def handle_combat(self):
-        # TODO: Very, very dumb combat.
         _tap_confirm()
 
-    # TODO: Render combat state
+    def execute(self, delta: float) -> bool:
+        # Update memory
+        active = self.update_mem()
+        # Handle FSM
+        match self.state:
+            case self._BattleFSM.PRE_BATTLE:
+                if active and not self.mem.ended:
+                    self.state = self._BattleFSM.BATTLE
+                    logger.debug("Pre battle => Battle")
+                else:
+                    # Tap confirm until battle starts
+                    _tap_confirm()
+                    return False
+            case self._BattleFSM.BATTLE:
+                if active:
+                    self.handle_combat()
+                    if self.mem.ended:
+                        logger.debug("Battle => Post battle")
+                        self.state = self._BattleFSM.POST_BATTLE
+                return False
+            case self._BattleFSM.POST_BATTLE:
+                # Tap past win screen/cutscenes
+                _tap_confirm()
+
+        return not active
+
     def render(self, window: WindowLayout) -> None:
         if not self.active:
             return
         window.stats.erase()
+        # Render header
         window.stats.write_centered(line=1, text="Evoland 1 TAS")
         window.stats.write_centered(line=2, text="ATB Combat")
-
+        # Render party and enemy stats
         window.stats.addstr(Vec2(1, 4), "Party:")
         self._print_group(window=window, group=self.mem.allies, y_offset=5)
         window.stats.addstr(Vec2(1, 8), "Enemies:")
         self._print_group(window=window, group=self.mem.enemies, y_offset=9)
-
-        # TODO: Hit/Damage prediction
-        # TODO: Who is acting?
+        # Render damage predictions
         if not self.mem.ended:
-            # TODO: Damage prediction for Kaeris
-            dmg = self.predict_damage(self.mem.allies[0], self.mem.enemies[0])
-            window.stats.addstr(Vec2(1, 13), "Damage prediction:")
-            window.stats.addstr(Vec2(2, 14), f"Clink: {dmg}")
-
-        # TODO: map representation
+            self._render_combat_predictions(window=window)
+            # TODO: map representation?
+
+    def _render_combat_predictions(self, window: WindowLayout):
```

Comment:
Added rendering attack prediction, based on currently active character and the first enemy.